net/interfaces/interfaces_darwin.go

```diff
@@ -17,3 +22,70 @@ func fetchRoutingTable() (rib []byte, err error) {
 func parseRoutingTable(rib []byte) ([]route.Message, error) {
 	return route.ParseRIB(syscall.NET_RT_IFLIST2, rib)
 }
+
+var (
+	mu      sync.Mutex
+	ifNames = make(map[int]string)
+)
+
+func getDelegatedInterface(ifIndex int) (int, error) {
+	mu.Lock()
+	defer mu.Unlock()
+
+	// To get the delegated interface, we do what ifconfig does and use the
+	// SIOCGIFDELEGATE ioctl. It operates in term of a ifreq struct, which
+	// has to be populated with a interface name. To avoid having to do a
+	// interface index -> name lookup every time, we cache interface names
+	// (since indexes and names are stable after boot).
+	ifName, ok := ifNames[ifIndex]
+	if !ok {
+		iface, err := net.InterfaceByIndex(ifIndex)
+		if err != nil {
+			return 0, err
+		}
+		ifName = iface.Name
+		ifNames[ifIndex] = ifName
+	}
+
+	// Only tunnels (like Tailscale itself) have a delegated interface, avoid
+	// the ioctl if we can.
+	if !strings.HasPrefix(ifName, "utun") {
+		return 0, nil
```

Comment:
is 0 ever used as a valid ifindex? if so, should this be -1 for none?

Reply:
0 is documented as never being used, e.g. from the `net.Interface` [godoc](https://pkg.go.dev/net@go1.20#Interface):

```go
Index        int          // positive integer that starts at one, zero is never used
```

Will document the 0 return value.